### ISSUE TITLE
installer re-downloads tarball every time

### ISSUE THREAD

**Issue by rpglover64** [NixOS/nix#2119]
Steps:

1. download install script from `https://nixos.org/nix/install`; (sha1sum = 92749879a8ce4a640ba8ae750f7f5555a9343711)
2. run `sh install`
3. have a failure after the tarball is downloaded (e.g. ctrl-c)
4. run `sh install` again

Expected:

old tarball is reused

Got:

tarball is re-downloaded

@shlevy  told me to mention @grahamc

**Comment by matthewbauer** [comment]
So there are some legitimate reasons to do this:

- We aren't sure what failed- it's possible the tarball is corrupted or not a real tarball.
- It's a fairly small download (IIRC <20MB).

I think the best behavior would be to start bundling the tarball into ```nixos.org/nix/install``` so that you can reuse install with something like Arx. But this is more of an issue with nixos-homepage, because you can always perform an install directly like this:

```
case "$(uname -s).$(uname -m)" in
    Linux.x86_64) system=x86_64-linux;;
    Linux.i?86) system=i686-linux;;
    Linux.aarch64) system=aarch64-linux;;
    Darwin.x86_64) system=x86_64-darwin;;
    *) exit 1;;
esac
version=2.0.1
name=nix-$version-$system

curl -O https://nixos.org/releases/nix/nix-$version/$name.tar.bz2
bzcat $name.tar.bz2  | tar -xf -
sh $name/install
```

**Comment by rpglover64** [comment]
> So there are some legitimate reasons to do this

Yes. In my particular use case, it caused problems.

> We aren't sure what failed- it's possible the tarball is corrupted or not a real tarball.

We could (and probably should) have a checksum in the script, and not try redownloading if the tarball matches the checksum.

> It's a fairly small download (IIRC <20MB).

It took me on the order of 30 seconds each time. Not a big deal, except that I ended up having to run the script around 20 times when it failed almost at the end of the installation and I needed to do some manual cleanup and debugging and try again.

> because you can always perform an install directly

I did end up changing the install script to look for a manually fetched tarball, but it was a pretty unpleasant user experience. If I had not had someone to help me and a specific need, I might have given up trying to install it.

**Comment by grahamc** [comment]
> We could (and probably should) have a checksum in the script, and not try redownloading if the tarball matches the checksum.

We already have the checksum in the script, and validate the hash of the tarball matches at install time. I'm not really clear what the bundling gets us?

**Comment by grahamc** [comment]
> How to verify it though? 

That is what this code did: https://github.com/NixOS/nixos-homepage/pull/215/files#diff-9f5a78be31376c52fce3efcbbb8c764bL44

**Comment by matthewbauer** [comment]
Yes, so to be clear the bundling does not exactly solve this issue. It more just solves the common case where the user has a downloaded "install" script and is surprised that every time they run it they need to download a new tarball.

**Comment by peti** [comment]
IMHO, the install script should *not* keep the downloaded install tarball around. What the script should do is *succeed* at the first attempt. Adding complexity to the script that's designed to make life easier for people who run it 20 times in a row feels like we're going into the wrong direction.

**Comment by rpglover64** [comment]
> What the script should do is succeed at the first attempt.

That would be ideal, sure; however, it is not _currently_ designed for that. For example, if there is any evidence of a previous nix installation, it fails with instructions on how to clean it (instructions which are incomplete, as I learned in #2120)

An alternative solution that would have worked for me is to provide (in addition to the `curl | sh` instructions) "advanced" instructions, for installing from the tarball, in a link in the "Get Nix" popup.

**Comment by nanbiz** [comment]
I have this issue when using a static nix binary and then calling for example `nix shell`. Is this related?
I don't know if this is relevant, but I have set the following in my config:
```
experimental-features = nix-command flakes
ssl-cert-file = /etc/pki/tls/certs/ca-bundle.crt
store = /tmp/myuser/store
```
(`/tmp` is never reset)
